pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,459 @@
+// impact performance as we iterate over all of them for some messages.
+//
+// TODO(baptist): Bump the default to a reasonable value like 10 that balances
+// usefulness with performance impact once we have validated the performance
+// impact.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemOnly,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces, 0 will disable tracing",
+	0,
+	settings.IntInRange(0, 1000),
+)
+
+// reserveSpace checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The soft invariant
+// is that numRegisteredStore <= numAllowed which can be temporarily violated if
+// MaxConcurrentRaftTraces is lowered. This method will return true if we can
+// add one to the number registered for both the store and replica, otherwise it
+// will return false. This method is optimized for the `numAllowed == 0` case
+// and avoids loading `numRegisteredStore` until after this check.`
+func (r *RaftTracer) reserveSpace() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegisteredReplica := r.numRegisteredReplica.Load()
+
+	// This can only occur if the numAllowed setting has changed since a
+	// previous call to reserveSpace. If this happens flush all our current
+	// traces and don't register this request. Note that when this happens we
+	// also wont't log this request.
+	if numRegisteredReplica > numAllowed {
+		log.Infof(r.ctx, "flushing all traces due to setting change")
+		r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+			r.removeEntry(index)
+			return true
+		})
+		return false
+	}
+
+	if numAllowed == 0 {
+		return false
+	}
+
+	// The maximum number of traces has been reached for the store. We don't
+	// register tracing and return false.
+	numRegisteredStore := r.numRegisteredStore.Load()
+	if numRegisteredStore >= numAllowed {
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	registerSucceeded := r.numRegisteredStore.CompareAndSwap(numRegisteredStore, numRegisteredStore+1)
+	if registerSucceeded {
+		// Add one unconditionally to the replica count.
+		r.numRegisteredReplica.Add(1)
+	}
+	// Note we can't assert numRegisteredStore <= numAllowed because if the
+	// setting is changed it can be temporarily violated on other replicas.
+	return registerSucceeded
+}
+
+// storeEntry stores the TracedEntry without a tracing context and returns the
+// newly created traceValue.
+func (r *RaftTracer) storeEntry(te kvserverpb.TracedEntry) *traceValue {
+	tv := traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	r.m.Store(te.Index, &tv)
```

Comment:
How does this behave if this index is already traced? Should we fail the insert? Right now it will override the previous item, right? But we've already increased the counters, so I'm not sure we won't leak the counter.

This can happen if the same entry index is passing by multiple times. E.g. the leader can retry sending `MsgApp`s, or two different leaders can send two different entries under the same index.

Reply:
Good catch! I added handling for this case. Specifically it doesn't register this but makes it clear what happened.